File: moviejava/app/src/main/java/com/example/moviejava/ui/home/fragment/now_playing/NowPlayingFragment.java
package com.example.moviejava.ui.home.fragment.now_playing;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.moviejava.databinding.FragmentNowPlayingBinding;
import com.example.moviejava.ui.base.BaseFragment;
import com.example.moviejava.ui.film.FilmActivity;
import com.example.moviejava.ui.home.HomeViewModel;
import com.example.moviejava.ui.home.adapter.MovieAdapter;

import java.util.Objects;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;


@AndroidEntryPoint
public class NowPlayingFragment extends BaseFragment<FragmentNowPlayingBinding> {

    private HomeViewModel viewModel;

    @Inject
    MovieAdapter viewAdapter;
    private int page = 1;

//    public NowPlayingFragment(@LayoutRes int contentLayoutId) {
//        super(contentLayoutId);
//    }

    @Override
    protected FragmentNowPlayingBinding onHandleViewBinding(LayoutInflater inflater, ViewGroup container) {
        return FragmentNowPlayingBinding.inflate(inflater, container, false);
    }

    @Override
    protected void onHandleObject(FragmentNowPlayingBinding rootView) {
        viewModel = new ViewModelProvider(requireActivity()).get(HomeViewModel.class);
        dataOnChange();
        setViewAdapter();
    }

    @Override
    protected void onHandleEvents() {
        setEvent();
    }

    private void dataOnChange() {
        defaultObserve(viewModel.nowPlaying,null,()->{
                    binding().loading.setVisibility(View.GONE);
                    binding().rcvNowPlaying.setVisibility(View.VISIBLE);
                }, isLoading -> {
                    if (isLoading)
                        binding().loading.setVisibility(View.VISIBLE);
                },
                it->viewAdapter.submitData(it));
    }

    private void setViewAdapter() {
        binding().rcvNowPlaying.setAdapter(viewAdapter);
        binding().rcvNowPlaying.setLayoutManager(new GridLayoutManager(getContext(), 4));
        binding().rcvNowPlaying.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                GridLayoutManager layoutManager = (GridLayoutManager) recyclerView.getLayoutManager();
                if (layoutManager!=null &&
                        layoutManager.findLastCompletelyVisibleItemPosition() == Objects.requireNonNull(recyclerView.getAdapter()).getItemCount() - 1) {
                    page++;
                    viewModel.getUpComing(page);
                }
            }
        });
    }

    private void setEvent() {
        viewAdapter.onClick = results -> {
            Intent intent = new Intent(getActivity(), FilmActivity.class);
            intent.putExtra("film_id", results.id);
            startActivity(intent);
        };

        binding().rcvNowPlaying.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                GridLayoutManager layoutManager = (GridLayoutManager) recyclerView.getLayoutManager();
                if (layoutManager != null &&
                        layoutManager.findLastCompletelyVisibleItemPosition() == Objects.requireNonNull(binding().rcvNowPlaying.getAdapter()).getItemCount() - 1) {
                    page++;
                    viewModel.getNowPlaying(page);
                }
            }
        });
    }


}

File: movie/app/src/main/java/com/example/movie/ui/home/fragment/now_playing/NowPlayingFragment.kt
package com.example.movie.ui.home.fragment.now_playing

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.example.movie.databinding.FragmentNowPlayingBinding
import com.example.movie.ui.base.BaseFragment
import com.example.movie.ui.film.FilmActivity
import com.example.movie.ui.home.HomeViewModel
import com.example.movie.ui.home.adapter.MovieAdapter
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class NowPlayingFragment : BaseFragment<FragmentNowPlayingBinding>() {
    @Inject
    lateinit var popularAdapter: MovieAdapter
    private var page = 1
    private val viewModel: HomeViewModel by activityViewModels()

    override fun setViewBinding(
        inflater: LayoutInflater, container: ViewGroup?
    ): FragmentNowPlayingBinding = FragmentNowPlayingBinding.inflate(inflater, container, false)

    override fun setupObjects(rootView: ViewBinding?) {
        binding.rcvNowPlaying.apply {
            adapter = popularAdapter
            layoutManager = GridLayoutManager(context, 4)
        }
        viewModel.apply {
            defaultObserver(nowPlaying, {
                binding.loading.visibility = View.GONE
                binding.rcvNowPlaying.visibility = View.VISIBLE
            }, {
                binding.loading.visibility = View.VISIBLE
               }, {
                it?.let { popularAdapter.submitData(it) }
            })
        }
    }

    override fun setupEvents() {
        popularAdapter.onClickItem = { r ->
            activity?.let {
                val intent = Intent(it, FilmActivity::class.java)
                intent.putExtra("id", r.id)
                it.startActivity(intent)
            }

        }
        binding.rcvNowPlaying.apply {
            addOnScrollListener(object : RecyclerView.OnScrollListener() {

                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    val layoutManager = recyclerView.layoutManager as GridLayoutManager
                    if (layoutManager.findLastCompletelyVisibleItemPosition() == adapter?.itemCount?.minus(
                            1
                        )
                    ) {
                        page++
                        viewModel.getPopular(page)
                    }
                }
            })
        }
    }
}